fold/fold.go

```diff
@@ -0,0 +1,879 @@
+	multibranchEnergy := a + b*float64(len(branches)) + c*float64(unpaired)
+
+	if unpaired == 0 {
+		multibranchEnergy = a + d
+	}
+
+	// energy of min-energy neighbors
+	e := multibranchEnergy + summedEnergy
+
+	// pointer to next structures
+	if helix {
+		// branches.pop()
+		branches = branches[:len(branches)-1]
+	}
+
+	return nucleicAcidStructure{energy: e, description: fmt.Sprintf("BIFURCATION:%dn/%dh", unpaired, branchCount), inner: branches}, nil
+}
+
+// internalLoop calculates the free energy of an internal loop.
+//
+// The first and last bp of both left and right sequences
+// are not themselves parts of the loop, but are the terminal
+// bp on either side of it. They are needed for when there's
+// a single internal looping bp (where just the mismatching
+// free energies are used)
+// Note that both left and right sequences are in 5' to 3' direction
+// This is adapted from the "Internal Loops" section of SantaLucia/Hicks, 2004
+// Args:
+//
+//		start:  The index of the start of structure on left side
+//		i1: The index to the right of start
+//		end:  The index of the end of structure on right side
+//		j1: The index to the left of end
+//	 foldContext: The FoldingContext for this sequence
+//
+// Returns the free energy associated with the internal loop
+func internalLoop(start, i1, end, j1 int, foldContext context) (float64, error) {
+	loopLeftIndex := i1 - start - 1
+	loopRightIndex := end - j1 - 1
+	loopLength := loopLeftIndex + loopRightIndex
+
+	if loopLeftIndex < 1 {
+		return 0, fmt.Errorf("internal loop: subsequence (%d, %d, %d, %d): missing left part of the loop", start, i1, end, j1)
+
+	}
+	if loopRightIndex < 1 {
+		return 0, fmt.Errorf("internal loop: subsequence (%d, %d, %d, %d): missing right part of the loop", start, i1, end, j1)
+	}
+
+	// single bp mismatch, sum up the two single mismatch pairs
+	if loopLeftIndex == 1 && loopRightIndex == 1 {
+		mismatchLeftEnergy := stack(start, i1, end, j1, foldContext)
+		mismatchRightEnergy := stack(i1-1, i1, j1+1, j1, foldContext)
+		return mismatchLeftEnergy + mismatchRightEnergy, nil
+	}
+	var enthalpyHDifference, entropySDifference, dG float64
+	// apply a penalty based on loop size
+	if foldEnergy, ok := foldContext.energies.internalLoops[loopLength]; ok {
+		enthalpyHDifference, entropySDifference = foldEnergy.enthalpyH, foldEnergy.entropyS
+		dG = deltaG(enthalpyHDifference, entropySDifference, foldContext.temp)
+	} else {
+		// it's too large an internal loop, extrapolate
+		foldEnergy := foldContext.energies.internalLoops[maxLenPreCalulated]
+		enthalpyHDifference, entropySDifference = foldEnergy.enthalpyH, foldEnergy.entropyS
+		dG = deltaG(enthalpyHDifference, entropySDifference, foldContext.temp)
+		dG = jacobsonStockmayer(loopLength, maxLenPreCalulated, dG, foldContext.temp)
+	}
+
+	// apply an asymmetry penalty
+	loopAsymmetry := math.Abs(float64(loopLeftIndex - loopRightIndex))
+	dG += 0.3 * loopAsymmetry
```

Comment:
what is 0.3?

Reply:
Or even from Santaluia 2004:
![image](https://user-images.githubusercontent.com/986630/232836903-e3d560a9-154c-4db5-aaa9-f6216b86e89f.png)